### DISCUSSION TITLE
While I extract images, if the images has no background color, the background would be black.

### DISCUSSION THREAD

**Discussion by w8741906** [pymupdf/PyMuPDF#2429] [Q&A]
_**Please provide all mandatory information!**_

## Describe the bug (mandatory)
While I extract images, if the images has no background color, the background would be black.

## To Reproduce (mandatory)
Explain the steps to reproduce the behavior, For example, include a minimal code snippet, example files, etc.

Here is my code.
```

def extract_images(pdf_path, output_folder, output_folder_with_title, classified_titles):
    doc = fitz.open(pdf_path)
    doc.colorspace = fitz.csRGB

    for page_number in range(len(doc)):
        page = doc[page_number]
        # only test 1 page
        # if page_number != 53:
        #     continue
        img_num = 0
        # get xref and rect objects
        images = page.get_images(full=True)
        for img_info in images:
            xref = img_info[0]
            img_rect = page.get_image_rects(xref)
            img_title = get_closest_title(page_number, img_rect, classified_titles)
            base_image = doc.extract_image(xref)
            image_bytes = base_image["image"]
            image = Image.open(io.BytesIO(image_bytes))
            img_num += 1
            image_name = "page" + str(page_number) + "_" + str(img_num) + img_title
            image_path = os.path.join(output_folder, f"{image_name}.png")
            image_with_title_path = os.path.join(output_folder_with_title, f"{image_name}.png")

            image.save(image_path)

            image_with_title = add_title_big(image, "page" + str(page_number) + "_" + str(img_num))
            image_with_title.save(image_with_title_path)

    return
```

## Expected behavior (optional)
extract a image with white background.

## Screenshots (optional)
This is what I extracted.
![22222](https://github.com/pymupdf/PyMuPDF/assets/16871641/55b78097-1beb-42f6-9643-f4674e345e15)

This is when I open it with adobe.
<img width="895" alt="1111" src="https://github.com/pymupdf/PyMuPDF/assets/16871641/dcec776c-02e1-4edd-b491-11bd71c0f216">

## Your configuration (mandatory)
 -  Mac book
 - Python 3.7
 - PyMuPDF version 1.22.1

**Comment by JorjMcKie** [comment]
This no bug, but a misconception. Transferring to "Discussions".

**Reply by JorjMcKie** [reply]
As mentioned in the documentation, `page.get_images()` also returns mask cross reference numbers. If that number is positive, you must combine the base image and the mask image to recreate the original transparent image.

**Reply by rafayaar** [reply]
@JorjMcKie Can you give a bit more clarity using some code example. Documentation is this case is a bit not clear to understand how we can combine base image with mask image.
Intuition is to extract images from PDF without background.

**Reply by JorjMcKie** [reply]
When an image with transparency (i.e. alpha channel) is inserted in a PDF, then the alpha channel is separated from the image and both resulting sub-images are stored as separate (XREF-ed) objects. This is a PDF-specific approach and has nothing to do with any tool (like (Py-) MuPDF).
To recover the original on image extraction, we therefore must identify and extract both pieces and combine them again.
If an image was transparent originally, the item in the `page.get_images()` list will have two positive integers as sub-items 0 and 1.

**Reply by JorjMcKie** [reply]
Item 1 is the xref of the alpha channel part.

**Reply by JorjMcKie** [reply]
So the easiest way is this:
```pymupdf
# given an item (basexref, maskxref, ...) in list page.get_images():
base = pymupdf.Pixmap(doc, basexref)
mask = pymupdf.Pixmap(doc, maskxref)
pix = pymupdf.Pixmap(base, mask)
```

**Reply by brightlightkim** [reply]
Perfect! That worked! Using Streamlit, I'm using this code to extract pdf files:

```
import streamlit as st
import fitz
from io import BytesIO
from PIL import Image

pdf = st.file_uploader("PDF here")

if not pdf:
    st.warning("No PDF file is uploaded.")
    st.stop()
        
doc = fitz.open(stream=pdf.read(), filetype="pdf")
    
for i in range(len(doc)):
    st.markdown(f"## Page {i+1}")
    for img_index, img in enumerate(doc.get_page_images(i)):
        base = fitz.Pixmap(doc, img[0])
        if img[1]:  # Check if there is a mask
            mask = fitz.Pixmap(doc, img[1])
            pix = fitz.Pixmap(base, mask)
        else:
            pix = base
        
        # Convert Pixmap to PIL Image for Streamlit
        image_data = BytesIO(pix.tobytes("png"))
        img_pil = Image.open(image_data)
        
        # Display the image
        st.image(img_pil, caption=f"Page {i + 1} Image {img_index + 1}")
        
    # Text extraction
    page = doc[i]
    text = page.get_text()
    st.write(text)
```
